ENIGMAsystem/SHELL/Platforms/Win32/WINDOWSfilemanip.cpp

```diff
@@ -107,92 +114,454 @@ void ini_section_delete(std::string section)
 
 /* OS Specific; should be moved */
 
-int file_exists(std::string fname) {
-    DWORD attributes = GetFileAttributes(fname.c_str());
-    if(attributes == 0xFFFFFFFF) {
-        return 0;
-    } else {
-        return 1;
-    }
+int file_exists(std::string fname)
+{
+    std::string str(fname);
+    std::replace(str.begin(),str.end(),'/','\\');
+
+    struct stat sb;
+    return (stat((char *)str.c_str(),&sb) == 0 &&
+        S_ISREG(sb.st_mode));
 }
 
-int file_delete(std::string fname) {
-    DWORD result = DeleteFileA(fname.c_str());
-
-    switch(result) {
-        case 0:
-            return 0;
-            break;
-        case ERROR_FILE_NOT_FOUND:
-            return 0;
-            break;
-        case ERROR_ACCESS_DENIED:
-            return 0;
-            break;
-        default:
-            return 1;
-            break;
+int file_delete(std::string fname)
+{
+    std::string str(fname);
+    std::replace(str.begin(),str.end(),'/','\\');
+
+    struct stat sb;
+    if (stat((char *)str.c_str(),&sb) == 0 &&
+        S_ISREG(sb.st_mode))
+    {
+        return (remove((char *)str.c_str()) == 0);
     }
+
+    return 0;
 }
 
-int file_rename(std::string oldname, std::string newname) {
-    DWORD result = MoveFileA(oldname.c_str(), newname.c_str());
+int file_rename(std::string oldname,std::string newname)
+{
+    std::string str1(oldname);
+    std::string str2(newname);
+    std::replace(str1.begin(),str1.end(),'/','\\');
+    std::replace(str2.begin(),str2.end(),'/','\\');
+
+    char rpath1[MAX_PATH];
+    char rpath2[MAX_PATH];
+    GetFullPathName((char *)str1.c_str(),MAX_PATH,rpath1,NULL);
+    GetFullPathName((char *)str2.c_str(),MAX_PATH,rpath2,NULL);
+    str1 = std::string(rpath1);
+    str2 = std::string(rpath2);
+
+    size_t slash1;
+    size_t slash2;
+    std::string dir1(str1);
+    std::string dir2(str2);
+
+    if (dir1.find('\\') != std::string::npos)
+        slash1 = dir1.find_last_of("\\");
+    else
+    {
+        dir1 = std::string("");
+        slash1 = 0;
+    }
+
+    if (dir2.find('\\') != std::string::npos)
+        slash2 = dir2.find_last_of("\\");
+    else
+    {
+        dir2 = std::string("");
+        slash2 = 0;
+    }
 
-    switch(result) {
-        case 0:
-            return 0;
-            break;
-        default:
-            return 1;
-            break;
+    int proceed = (dir1.substr(0,slash1) != dir2.substr(0,slash2));
+
+    if (proceed || basename((char *)str1.c_str()) != basename((char *)str2.c_str()))
+    {
+        struct stat sb1;
+        struct stat sb2;
+        if (stat((char *)str1.c_str(),&sb1) == 0 &&
+            S_ISREG(sb1.st_mode) &&
+            stat((char *)str2.c_str(),&sb2) != 0)
+        {
+            return (rename((char *)str1.c_str(),(char *)str2.c_str()) == 0);
```

Comment:
Same comments regarding `stat`/`rename`. They're just not Windows functions. Same comments regarding `MoveFileA` vs `MoveFileW`, if unicode support is what you're after.